### PR TITLE
Matching

### PR DESCRIPTION
**Pull request by MAlberts99**
The PR adds the following functionality to the NMR Matching data generation script:

- Creating randomly drawn sets of molecules from the data. The number of molecules per set is varied in a range from 2-7.
- Create Sets with a given Tanimoto similarity, i.e. the similarity of all molecules in the set falls in an Interval of Tanimoto similarities
- Sets where there is no match between the set and the spectra

### CODE REVIEW

File: src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py [hunk 1]
```diff
@@ -337,7 +337,7 @@ def main(
     # Load dataframe containing nmr data
     logger.info("Reading data.")
     nmr_df = pd.read_pickle(nmr_data)
-    nmr_df.drop(columns=["1H_NMR_exp", "13C_NMR_exp"], inplace=True)
+    #nmr_df.drop(columns=["1H_NMR_exp", "13C_NMR_exp"], inplace=True)
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py]: If not necessary anymore: delete it. Otherwise, you can specify why it is commented out as a comment on the line before

File: src/nmr_to_structure/prepare_input/nmr_utils.py [hunk 2]
```diff
@@ -1,44 +1,98 @@
...
+from rdkit import Chem, DataStructs
+from rdkit.Chem import AllChem, rdMolDescriptors
 from rxn.chemutils.tokenization import tokenize_smiles
+from rdkit import RDLogger
 from sklearn.model_selection import train_test_split
 
-RANDOM_SEED = 3246
+DEFAULT_SEED = 3246
+DEFAULT_NON_MATCHING_TOKEN = "<no_match> <no_match> <no_match> <no_match> <no_match> <no_match> <no_match> <no_match> <no_match> <no_match>"
+DEFAULT_ALLOWED_ELEMENTS = set(["C", "H", "N", "O", "S", "P", "F", "Cl", "Br", "I"])
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/nmr_utils.py:21]:
```suggestion
DEFAULT_ALLOWED_ELEMENTS = {"C", "H", "N", "O", "S", "P", "F", "Cl", "Br", "I"}
```
slightly more efficient

File: src/nmr_to_structure/prepare_input/nmr_utils.py [hunk 3]
```diff
@@ -1,44 +1,98 @@
...
-def jitter(value, jitter_range: float = 2):
+def jitter(value: float, jitter_range: float = 2) -> float:
     jitter_value = np.random.uniform(-jitter_range, +jitter_range)
     return value + jitter_value
 
 
-def split_data(input_data: Any) -> Tuple[Any, Any, Any]:
+def split_data(
+    input_data: Any, test_size: float = 0.1, val_size: float = 0.05
+) -> Tuple[Any, Any, Any]:
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/nmr_utils.py:36]: If you can: avoid `Any`. Would it be `pd.DataFrame` here?

File: src/nmr_to_structure/prepare_input/nmr_utils.py [hunk 4]
```diff
@@ -1,44 +1,98 @@
...
     train_data, test_data = train_test_split(
-        input_data, test_size=0.1, random_state=RANDOM_SEED
+        input_data, test_size=test_size, random_state=DEFAULT_SEED
     )
     train_data, val_data = train_test_split(
-        train_data, test_size=0.05, random_state=RANDOM_SEED
+        train_data, test_size=val_size, random_state=DEFAULT_SEED
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/nmr_utils.py:41]: to make your function slightly more reusable: you can add an argument for the random seed to the function, `, random_seed=DEFAULT_SEED) -> ...`

File: src/nmr_to_structure/prepare_input/nmr_utils.py [hunk 5]
```diff
@@ -1,44 +1,98 @@
...
     )
 
     return (train_data, test_data, val_data)
 
 
+def evaluate_molecule(smiles: str) -> bool:
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/nmr_utils.py:47]: I would add a short docstring to clarify what it means to "evaluate" a molecule, and what True/False as a return value would mean.

File: src/nmr_to_structure/prepare_input/nmr_utils.py [hunk 6]
```diff
@@ -1,44 +1,98 @@
...
+    RDLogger.DisableLog('rdApp.*')
+    mol = Chem.MolFromSmiles(smiles)
+    if mol is None:
+        return False
+
+    formula = rdMolDescriptors.CalcMolFormula(mol)
+    atoms = re.findall(r"[A-Z][a-z]*\d*", formula)
+    atoms_clean = set([re.sub(r"\d+", "", atom) for atom in atoms])
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/nmr_utils.py:55]:
```suggestion
    atoms_clean = {re.sub(r"\d+", "", atom) for atom in atoms}
```

File: src/nmr_to_structure/prepare_input/nmr_utils.py [hunk 7]
```diff
@@ -1,44 +1,98 @@
...
+        for item in smiles:
+            f.write(f"{item}\n")
+
+    nmr_input = data_set.nmr_input
+    with open(out_path / f"src-{set_type}.txt", "w") as f:
+        for item in nmr_input:
+            f.write(f"{item}\n")
+
+
+# Copied from github.com/rxn4chemistry/rxn-onmt-models/blob/main/src/rxn/onmt_models/utils.py
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/nmr_utils.py:89]: I would just move this to make it part of the docstring of the function.

File: src/nmr_to_structure/prepare_input/nmr_utils.py [hunk 8]
```diff
@@ -1,44 +1,98 @@
...
+        return now_formatted + ".log"
+    else:
+        return prefix + "-" + now_formatted + ".log"
+
+
+# Functions for making NMR strings #
 def build_1H_peak(
-    HNMR_sim_peaks,
-    peak,
+    HNMR_sim_peaks: dict,
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/nmr_utils.py:109]: for the annotation: Dict of what to what? -> f.i. `Dict[str, Dicst[str, float]]` or so

File: src/nmr_to_structure/prepare_input/nmr_utils.py [hunk 9]
```diff
@@ -179,17 +233,215 @@ def build_cnmr_string(
...
 
 
-def save_set(data_set: pd.DataFrame, out_path: Path, set_type: str) -> None:
-    smiles = list(data_set.smiles)
-    smiles = [tokenize_smiles(smile) for smile in smiles]
+def make_nmr(
+    mode: str,
+    component: str,
+    hnmr: Optional[dict] = None,
+    cnmr: Optional[dict] = None,
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/nmr_utils.py:258]: dict of what to what?

File: src/nmr_to_structure/prepare_input/nmr_utils.py [hunk 10]
```diff
@@ -179,17 +233,215 @@ def build_cnmr_string(
...
+    return nmr_string
+
+
+# Functions and classes supporting the generation of Tanimoto sets #
+class TanimotoContainer:
+    def __init__(self, n_tanimoto_bins: int) -> None:
+        tanimoto_boxes = np.linspace(0, 1, n_tanimoto_bins + 1)
+
+        self.length = 0
+        self.tanimoto_results_container: Dict[str, list] = {
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/nmr_utils.py:316]: for the type annotation: `list` of what? -> `List[float]`?

File: src/nmr_to_structure/prepare_input/nmr_utils.py [hunk 11]
```diff
@@ -179,17 +233,215 @@ def build_cnmr_string(
...
+            "{:.3f}".format(tanimoto_boxes[i]): list() for i in range(n_tanimoto_bins)
+        }
+
+    def add(self, to_add_dict: Dict) -> None:
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/nmr_utils.py:320]: dict of what to what?

File: src/nmr_to_structure/prepare_input/nmr_utils.py [hunk 12]
```diff
@@ -179,17 +233,215 @@ def build_cnmr_string(
...
+        for key in to_add_dict:
+            self.tanimoto_results_container[key].extend(to_add_dict[key])
+            self.length += len(to_add_dict[key])
+
+    def __len__(self) -> int:
+        return self.length
+
+    def __getitem__(self, key: str) -> List:
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/nmr_utils.py:328]: `List[float]`?

File: src/nmr_to_structure/prepare_input/nmr_utils.py [hunk 13]
```diff
@@ -179,17 +233,215 @@ def build_cnmr_string(
...
+
+def get_tanimoto_set(
+    tgt_smiles: str,
+    nmr_df: pd.DataFrame,
+    mode: str,
+    hnmr_mode: str,
+    token_space: str,
+    non_matching: bool,
+    n_tanimoto_bins: int,
+) -> Dict[str, list]:
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/nmr_utils.py:370]: `List[float]`?

File: src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py [hunk 14]
```diff
@@ -1,92 +1,163 @@
...
+    save_set,
+    split_data,
+    tanimoto_set_worker,
+)
+
+DEFAULT_TEST_SIZE = 0.1
+DEFAULT_VAL_SIZE = 0.05
+DEFAULT_MAX_SEQ_LEN = 600
+
+random.seed(DEFAULT_SEED)
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py]: This one can have undesired side effect when someone imports this module. Instead, it is better to put this inside the main function.

File: src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py [hunk 15]
```diff
@@ -1,92 +1,163 @@
...
+    random.shuffle(all_smiles)
+
+    smiles_chunks = [
+        all_smiles[
+            i
+            * len(all_smiles)
+            // cores : min((i + 1) * len(all_smiles) // cores, len(all_smiles))
+        ]
+        for i in range(cores)
+    ]
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py:58]: For things like this one, I would typically try to use a function from another package, as it's easy to not spot a mistake. I see for instance that there is `numpy.array_split` for that purpose. Alternatively, I would put it in its own function (which has associated unit tests). See also here: https://stackoverflow.com/a/312464

File: src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py [hunk 16]
```diff
@@ -1,92 +1,163 @@
...
-            header=False,
-            token_space=token_space,
-        )[0]
-        nmr_string = " " + cnmr_string
-    
-    return nmr_string
+    return combined_df
 
 
 def make_nmr_rxn_set(
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py:106]: this function is quite long and a bit complex; if you were to refactor some of the code, this is a typical one that I would try to encapsulate in a class with small modular functions

File: src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py [hunk 17]
```diff
@@ -1,92 +1,163 @@
...
+                component,
+                hnmr=nmr_df.loc[component]["1H_NMR_sim"]["peaks"],
+                cnmr=nmr_df.loc[component]["13C_NMR_sim"],
+                hnmr_mode=hnmr_mode,
+                token_space=token_space,
+            )
+            nmr_input = components_tokenized + nmr
+
+            if nmr_input.count(" ") + 1 > DEFAULT_MAX_SEQ_LEN:
+                continue
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py:159]: Add a small comment to say (if I understand correctly) that too long sequences are ignored?

Depending on how often this happens, you may want to consider adding a `logger.info` or so

File: src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py [hunk 18]
```diff
@@ -1,92 +1,163 @@
...
 ) -> pd.DataFrame:
     rxns = list(rxn_set.index)
-    mols = set(nmr_set.index)
+    random.shuffle(rxns)
+
+    mols = set(nmr_df.index)
     mols_sample = tuple(mols)
 
-    src_tgt_pairs = list()
+    src_tgt_pairs: List[dict] = list()
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py:122]: Do you need the content to be a dictionary? If not, `List[Tuple[str, str]]` may be more appropriate.

File: src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py [hunk 19]
```diff
@@ -1,92 +1,163 @@
...
 
-    for rxn in tqdm(rxns):
-        rxn = rxn.replace(">", ".").replace("..", "")
+    idx = 0
+    pbar = tqdm(total=n_samples)
+    while len(src_tgt_pairs) < n_samples:
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py:126]: this `while` loop looks like something that wants to be a `for` loop instead. I am not sure if you actually need the `idx` variable.

You can do a `break` when the size reaches the desired one, and do the check for "Ran out of reactions to consider before reaching desired training samples" outside of the for loop.

File: src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py [hunk 20]
```diff
@@ -95,64 +166,108 @@ def make_nmr_rxn_set(
...
+    # Default mol_distribution to cover equal number of examples for the different set sizes, if non_matching is set to true -> divide by two as two examples will be added per pass
+    mol_distribution = [
+        [
+            i,
+            int(n_samples / ((n_max_mol_set_size - 1) * 2))
+            if non_matching
+            else int(n_samples / (n_max_mol_set_size - 1)),
+        ]
+        for i in range(2, n_max_mol_set_size + 1)
+    ]
```
Comment (avaucher) [inline src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py:222]: I don't really understand what is happening here. What does `mol_distribution` contain after that?

You can consider putting this in a separate function.

you can then probably put `int(n_samples / ((n_max_mol_set_size - 1) * 2)` in a separate variable before the for loop, with a descriptive name? It does not need to be evaluated at every iteration of the loop.

**Review by avaucher — COMMENTED** [review]
Thanks for the PR!

There's no real comment on things that have to change; it looks like everything is fine as it is. Most of the comments are rather as hints on things I would do differently in the future. 

For instance, for type annotations I would always use `List[SomeValue]` instead of `list`, `Dict[str, SomeValue]` instead of `dict`, etc. It makes it more readable, as people know directly what format is expected / returned by a function.

Also, in general, when your functions become quite long with several levels of indentations, it can be a sign that a class can be more adequate - this leads to smaller functions and usually makes the code more readable (if the function names are descriptive).
